frontend/src/main/scala/bloop/bsp/BloopBspServices.scala

```diff
@@ -0,0 +1,141 @@
+   */
+  private final val bspForwarderLogger = new bloop.logging.AbstractLogger() {
+    override def name: String = "bsp-logger"
+    override def verbose[T](op: => T): T = op
+    override def ansiCodesSupported(): Boolean = true
+
+    override def debug(msg: String): Unit = {
+      callSiteState.logger.debug(msg)
+      //Window.showMessage.info(msg)(client)
+    }
+
+    override def error(msg: String): Unit = {
+      callSiteState.logger.error(msg)
+      Window.showMessage.error(msg)(client)
+    }
+
+    override def warn(msg: String): Unit = {
+      callSiteState.logger.warn(msg)
+      Window.showMessage.warn(msg)(client)
+    }
+
+    override def trace(t: Throwable): Unit = {
+      Window.showMessage.info(t.toString)(client)
+    }
+
+    override def info(msg: String): Unit = {
+      callSiteState.logger.info(msg)
+      Window.showMessage.info(msg)(client)
+    }
+  }
+
+  // Internal state, think how to make this more elegant.
+  @volatile private final var currentState: State = null
+
+  /**
+   * Get the latest state that can be reused and cached by bloop so that the next client
+   * can have access to it.
+   *
+   * @return
+   */
+  def latestState: State = {
+    val state0 = if (currentState == null) callSiteState else currentState
+    state0.copy(logger = callSiteState.logger)
+  }
+
+  private final val defaultOpts = callSiteState.commonOptions
+  def reloadState(uri: String): State = {
+    val state0 = State.loadStateFor(AbsolutePath(uri), defaultOpts, bspForwarderLogger)
+    state0.copy(logger = bspForwarderLogger, commonOptions = latestState.commonOptions)
+  }
+
+  /**
+   * Implements the initialize method that is the first pass of the Client-Server handshake.
+   *
+   * @param initializeBuildParams The params request that we get from the client.
+   * @return An async computation that returns the response to the client.
+   */
+  def initialize(
+      initializeBuildParams: InitializeBuildParams
+  ): MonixTask[Either[JsonRpcResponse.Error, InitializeBuildResult]] = MonixTask {
+    callSiteState.logger.info("request received: build/initialize")
+    bspLogger.info("request received: build/initialize")
+    currentState = reloadState(initializeBuildParams.rootUri)
+    Right(
+      InitializeBuildResult(
+        Some(
+          BuildServerCapabilities(
+            compileProvider = true,
+            textDocumentBuildTargetsProvider = true,
+            dependencySourcesProvider = false,
+            buildTargetChangedProvider = true
+          )
+        )
+      )
+    )
+  }
+
+  def initialized(
+      initializedBuildParams: InitializedBuildParams
+  ): Unit = {
+    bspLogger.info("BSP initialization handshake complete.")
+  }
+
+  def compile(params: CompileParams): MonixTask[Either[JsonRpcResponse.Error, CompileReport]] = {
+    MonixTask {
+      // TODO(jvican): Naive approach, we need to implement batching here.
```

Comment:
Can you open an issue for that if you don't want to take care of it before we merge?

Reply:
Yes

Reply:
https://github.com/scalacenter/bloop/issues/217